### PR TITLE
RRTMGP refactoring

### PR DESCRIPTION
**Pull request by dustinswales**
This PR contains changes to the RRTMGP physics.

- Major refactoring of RRTMGP interstitials. collapse all rrtmgp_lw(sw) scheme files into rrtmgp_lw(sw)_main schemes. This reduces the number of Interstitial fields needed by the GP scheme and allows us to use the rte-rrtmgp submodule directly, w/o the need for a special branch containing the CCPP hooks.
- For longwave radiation calculation, add "blocking" capability to main loop.
- Fix bug in coupling of RRTMGP shortwave radiation to surface.
- CMake cleanup.

### CODE REVIEW

File: physics/GFS_rrtmgp_cloud_mp.F90 [hunk 1]
```diff
@@ -1,10 +1,5 @@
-!> \file GFS_rrtmgp_cloud_mp.F90
-!!
-!> \defgroup GFS_rrtmgp_cloud_mp GFS_rrtmgp_cloud_mp.F90
-!!
-!! \brief This module contains the interface for ALL cloud microphysics assumptions and 
-!! the RRTMGP radiation scheme. Specific details below in subroutines.
-!!
+! ########################################################################################
```
Comment (grantfirl) [inline physics/GFS_rrtmgp_cloud_mp.F90]: I'm curious why this Doxygen markup was removed since the contents of this file don't really change much.

File: physics/GFS_rrtmgp_post.F90 [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+         sfcdsw               ! SW sfc all-sky     downward flux   (W/m2)
+    real(kind_phys), dimension(:,:), intent(inout) :: &
+         htrlw,             & ! LW all-sky heating rate (K/s)
+         htrsw,             & ! SW all-sky heating rate (K/s)
+         htrlwu               ! LW all-sky heating-rate updated in-between radiation calls.
+    type(sfcflw_type), dimension(:), intent(inout) :: &
+         sfcflw               ! LW radiation fluxes at sfc
+    type(sfcfsw_type), dimension(:), intent(inout) :: &
+         sfcfsw               ! SW radiation fluxes at sfc
+    type(topfsw_type), dimension(:), intent(out) :: &
```
Comment (grantfirl) [inline physics/GFS_rrtmgp_post.F90]: I don't know if this will lead to problems or not, but topfsw and topflw are not set to anything if doLWrad and doSWrad are F since they are intent(out).

File: physics/GFS_rrtmgp_pre.F90 [hunk 3]
```diff
@@ -164,7 +166,8 @@ subroutine GFS_rrtmgp_pre_run(me, nCol, nLev, nTracers, i_o3, lsswr, lslwr, fhsw
     integer, intent(out) :: &
```
Comment (grantfirl) [inline physics/GFS_rrtmgp_pre.F90:166]: Same comment RE: intent(out). Since many of these routines immediatley return if doSWrad and doLWrad are F, any intent(out) variables will automatically be set to random garbage upon return (whatever was previously in that memory space). This may not matter if they're not used, but it is good practice (and in the CCPP rules somewhere?) to initialize all intent(out) variables in CCPP schemes.

File: physics/GFS_rrtmgp_pre.F90 [hunk 4]
```diff
@@ -15,16 +15,19 @@ module GFS_rrtmgp_pre
        NF_VGAS,                  & !< Number of active gas species
        getgases,                 & !< Routine to setup trace gases
        getozn                      !< Routine to setup ozone
-  ! RRTMGP types
-  use mo_gas_concentrations, only: ty_gas_concs
   use radiation_tools,       only: check_error_msg,cmp_tlev
+  use rrtmgp_lw_gas_optics,  only: lw_gas_props
```
Comment (grantfirl) [inline physics/GFS_rrtmgp_pre.F90:19]: @dustinswales This reminds me to ask about RRTMGP DDTs, since I've forgotten where this landed when it was implemented. In the past, I think that the we have said that if CCPP schemes use their own DDTs that they be confined to those that only used "internally" in physics. I.e., that in the long run we're not requiring hosts to "know" about more DDTs than necessary. So, in practice, that would mean confining RRTMGP-specific DDTs to the current Interstitial top-level DDT because all of those will technically be able to be allocated by the CCPP in the future with capgen. Is this your understanding and are your changes consistent with this? Am I totally off about this?

File: physics/GFS_rrtmgp_sw_pre.F90 [hunk 5]
```diff
@@ -1,95 +0,0 @@
...
-       idxday = 0
-       do i = 1, nCol
-          if (coszen(i) >= 0.0001) then
-             nday = nday + 1
-             idxday(nday) = i
-          endif
-       enddo
-
-       ! Spread across all SW bands
-       do iBand=1,sw_gas_props%get_nband()
```
Comment (grantfirl) [inline physics/GFS_rrtmgp_sw_pre.F90:78]: Were these lines (sfc_alb_X_Y_byband) moved somewhere else or just removed?

File: physics/rrtmgp_aerosol_optics.F90 [hunk 6]
```diff
@@ -65,34 +57,37 @@ subroutine rrtmgp_aerosol_optics_run(doSWrad, doLWrad, nCol, nLev, nTracer, nTra
     ! Outputs
     real(kind_phys), dimension(:,:), intent(out) :: &
          aerodp                   ! Vertical integrated optical depth for various aerosol species 
-    type(ty_optical_props_2str),intent(out) :: &
-         sw_optical_props_aerosol ! RRTMGP DDT: Longwave aerosol optical properties (tau)
-    type(ty_optical_props_1scl),intent(inout) :: &
-         lw_optical_props_aerosol ! RRTMGP DDT: Longwave aerosol optical properties (tau)
+    real(kind_phys), dimension(:,:,:), intent(out) :: &
```
Comment (grantfirl) [inline physics/rrtmgp_aerosol_optics.F90:60]: Same comment about uninitialized intent(out) variables. It could be fine; just need to doublecheck. I mean, it wasn't causing problems before, so why should it now?

File: physics/rrtmgp_aerosol_optics.F90 [hunk 7]
```diff
@@ -65,34 +57,37 @@ subroutine rrtmgp_aerosol_optics_run(doSWrad, doLWrad, nCol, nLev, nTracer, nTra
...
+         aersw_ssa,             & ! Shortwave aerosol single scattering albedo
+         aersw_g                  ! Shortwave aerosol asymmetry parameter
     integer, intent(out) :: &
          errflg                   ! CCPP error flag
     character(len=*), intent(out) :: &
          errmsg                   ! CCPP error message
 
     ! Local variables
-    real(kind_phys), dimension(nCol, nLev, lw_gas_props%get_nband(), NF_AELW) :: &
+    real(kind_phys), dimension(nCol, nLev, lw_gas_props%get_nband(), 3) :: &
```
Comment (grantfirl) [inline physics/rrtmgp_aerosol_optics.F90:73]: seems a little weird to be going from NF_AELW to a hard-coded constant. I guess it will never change?

File: physics/rrtmgp_lw_cloud_optics.F90 [hunk 8]
```diff
@@ -74,55 +72,42 @@ module rrtmgp_lw_cloud_optics
 
 contains
 
-!>\defgroup rrtmgp_lw_cloud_optics_mod GFS RRTMGP-LW Cloud Optics Module
```
Comment (grantfirl) [inline physics/rrtmgp_lw_cloud_optics.F90:77]: Same comment about removing Doxygen stuff. We may need to consult Man about this to make sure it doesn't mess up the scientific docs too much.

File: physics/rrtmgp_lw_cloud_optics.F90 [hunk 9]
```diff
@@ -391,171 +376,4 @@ subroutine rrtmgp_lw_cloud_optics_init(nrghice, mpicomm, mpirank, mpiroot,
...
-  ! ######################################################################################
-!> \section arg_table_rrtmgp_lw_cloud_optics_run
-!! \htmlinclude rrtmgp_lw_cloud_optics.html
-!!
-!> \ingroup rrtmgp_lw_cloud_optics
-!!
-!! Compute longwave optical prperties (optical-depth) for ALL cloud types visible to RRTMGP.                                                                                                                                   
-!!
-!! \section rrtmgp_lw_gas_optics_run
-  subroutine rrtmgp_lw_cloud_optics_run(doLWrad, doG_cldoptics, icliq_lw, icice_lw,      &
```
Comment (grantfirl) [inline physics/rrtmgp_lw_cloud_optics.F90:403]: I can't find where this code was moved to. Was it removed?

Comment (grantfirl) [reply physics/rrtmgp_lw_cloud_optics.F90:403]: rrtmgp_lw_main?

File: physics/rrtmgp_sw_main.F90 [hunk 10]
```diff
@@ -0,0 +1,682 @@
...
+         active_gases_array     ! List of active gases from namelist as array
+    real(kind_phys), intent(in) :: &
+         solcon                 ! Solar constant
+
+    ! Outputs
+    character(len=*), intent(out) :: &
+         errmsg                ! CCPP error message
+    integer, intent(out) :: &
+         errflg                ! CCPP error flag
+    real(kind_phys), dimension(:,:), intent(out) :: &
```
Comment (grantfirl) [inline physics/rrtmgp_sw_main.F90]: Same comment RE: intent(out)

**Comment by grantfirl** [comment]
@dustinswales I'm finished looking through this. Looks fine to me for the most part, with some worries about uninitialized intent(out) variables. It seems like every small scheme that was deleted ended up in the SW or LW main schemes? It's a little difficult to follow code moving between files in the GitHub interface (without looking really hard).

File: physics/GFS_rrtmgp_cloud_mp.F90 (hunk 1 above)
Comment (dustinswales) [reply physics/GFS_rrtmgp_cloud_mp.F90]: Fixed.

File: physics/GFS_rrtmgp_post.F90 (hunk 2 above)
Comment (dustinswales) [reply physics/GFS_rrtmgp_post.F90]: They aren't used anywhere not guarded by doLWrad and doSWrad, but you're correct. I cleaned this up

File: physics/GFS_rrtmgp_pre.F90 (hunk 4 above)
Comment (dustinswales) [reply physics/GFS_rrtmgp_pre.F90:19]: @grantfirl This PR address the concerns you brought up. Up until now, there were two types of DDTs in the GP scheme: internal DDTs that exist with out the model's knowledge (i.e. lw_gas_props), and interstitial DDT's. For the former we had to add hooks for ccpp compliance in the rte-rrtmgp submodule, and also inform the host about these GP DDTs. This was a headache to manage and is now longer needed with the refactoring.

File: physics/GFS_rrtmgp_sw_pre.F90 (hunk 5 above)
Comment (dustinswales) [reply physics/GFS_rrtmgp_sw_pre.F90:78]: They were moved to [rrtmgp_sw_main.F90](https://github.com/dustinswales/ccpp-physics/blob/rrtmgp_refactor/physics/rrtmgp_sw_main.F90#L359). This is also part of the shortwave surface coupling bugfix (there is a similar spectral correction done in the _post step).

File: physics/rrtmgp_lw_cloud_optics.F90 (hunk 9 above)
Comment (dustinswales) [reply physics/rrtmgp_lw_cloud_optics.F90:403]: Yes. The rrtmgp_lw_cloud_optics_run routine is now in [rtmgp_lw_main.F90](https://github.com/dustinswales/ccpp-physics/blob/rrtmgp_refactor/physics/rrtmgp_lw_main.F90#L362)
Similarly, the rrtmgp_lw_gas_optics_run routine is called just before the cloud-optics in the link above.

**Comment by dustinswales** [comment]
> @dustinswales I'm finished looking through this. Looks fine to me for the most part, with some worries about uninitialized intent(out) variables. It seems like every small scheme that was deleted ended up in the SW or LW main schemes? It's a little difficult to follow code moving between files in the GitHub interface (without looking really hard).

@grantfirl Thanks for your review. I addressed your concerns about the intent of some variables.

I admit this is hard to follow, my apologizes, it's a total restructuring of how GP is implemented in the CCPP.

File: physics/GFS_rrtmgp_cloud_mp.F90 [hunk 11]
```diff
@@ -484,47 +485,48 @@ subroutine cloud_mp_SAMF(nCol, nLev, t_lay, p_lev, p_lay, qs_lay, relhum,
...
+    tem0 = 1.0e5/con_g
     do iLay = 1, nLev
        do iCol = 1, nCol
           if (cnv_mixratio(iCol,iLay) > 0._kind_phys) then
              tem1   = min(1.0, max(0.0, (con_ttp-t_lay(iCol,iLay))*0.05))
              deltaP = abs(p_lev(iCol,iLay+1)-p_lev(iCol,iLay))*0.01
-             clwc   = max(0.0, cnv_mixratio(iCol,iLay)) * con_g * deltaP
-             cld_cnv_iwp(iCol,iLay) = clwc * tem1
-             cld_cnv_lwp(iCol,iLay) = clwc - cld_cnv_iwp(iCol,iLay)
+             clwc   = max(0.0, cnv_mixratio(iCol,iLay)) * tem0 * deltaP
```
Comment (ChunxiZhang-NOAA) [inline physics/GFS_rrtmgp_cloud_mp.F90:515]: You modified from con_g to 1.0e5/con_g. What is 1.0e5?

Comment (dustinswales) [reply physics/GFS_rrtmgp_cloud_mp.F90:515]: @ChunxiZhang-NOAA This was a bug fix identified by @Qingfu-Liu back in June.

Comment (ChunxiZhang-NOAA) [reply physics/GFS_rrtmgp_cloud_mp.F90:515]: @dustinswales Thanks for clarifying.

**Review by ChunxiZhang-NOAA — APPROVED** [review]

**Comment by jkbk2004** [comment]
We are about to start working on https://github.com/ufs-community/ufs-weather-model/pull/1571. @dustinswales can you mark out unsolved conversations?

File: physics/rrtmgp_aerosol_optics.F90 (hunk 7 above)
Comment (dustinswales) [reply physics/rrtmgp_aerosol_optics.F90:73]: @grantfirl In this context 3 is fixed. The third dimension indexes different fields in a stacked array, optical-depth, single-scattering albedo, and asymmetry parameter
